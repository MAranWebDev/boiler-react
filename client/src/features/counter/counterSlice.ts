/* redux */
import { type PayloadAction, createSlice } from '@reduxjs/toolkit';

const counterSlice = createSlice({
  name: 'counter',
  initialState: {
    counter: 0,
  },
  reducers: {
    increment(state, action: PayloadAction<number>) {
      state.counter += action.payload;
    },
    decrement(state, action: PayloadAction<number>) {
      state.counter -= action.payload;
    },
    reset(state, action: PayloadAction<number>) {
      state.counter = action.payload;
    },
  },
});

export const counterReducer = counterSlice.reducer;
export const { increment, decrement, reset } = counterSlice.actions;
