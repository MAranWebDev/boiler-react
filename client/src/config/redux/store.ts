/* redux */
import { configureStore } from '@reduxjs/toolkit';

/* features */
import { counterReducer } from '../../features/counter';

// Infer the `RootState` and `AppDispatch` types from the store itself
type RootState = ReturnType<typeof store.getState>;
type AppDispatch = typeof store.dispatch;

const store = configureStore({
  reducer: {
    counter: counterReducer,
  },
});

export { type RootState, type AppDispatch, store };
