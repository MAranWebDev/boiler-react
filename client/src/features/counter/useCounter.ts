/* react */
import { useEffect } from 'react';

/* redux */
import { useAppDispatch } from '../../config/redux/hooks';
import { increment, decrement, reset } from './counterSlice';

const useCounter = (initialValue: number) => {
  const dispatch = useAppDispatch();

  useEffect(() => {
    dispatch(reset(initialValue));
  }, [dispatch, initialValue]);

  const handleIncrement = () => dispatch(increment(1));
  const handleDecrement = () => dispatch(decrement(1));
  const handleReset = () => dispatch(reset(initialValue));

  return { handleIncrement, handleDecrement, handleReset };
};

export { useCounter };
