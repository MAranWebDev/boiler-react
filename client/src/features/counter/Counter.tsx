/* redux */
import { useAppSelector } from '../../config/redux/hooks';

/* hooks */
import { useCounter } from './useCounter';

interface CounterProps {
  value: number;
}

function Counter({ value }: CounterProps) {
  const { counter } = useAppSelector((state) => state.counter);
  const { handleIncrement, handleDecrement, handleReset } = useCounter(value);

  return (
    <div>
      <section>
        <h2>{counter}</h2>
      </section>

      <section>
        <button
          type="button"
          aria-label="btn-increment"
          onClick={handleIncrement}
        >
          +1
        </button>
        <button
          type="button"
          aria-label="btn-decrement"
          onClick={handleDecrement}
        >
          -1
        </button>
        <button type="button" aria-label="btn-reset" onClick={handleReset}>
          Reset
        </button>
      </section>
    </div>
  );
}

export { Counter };
