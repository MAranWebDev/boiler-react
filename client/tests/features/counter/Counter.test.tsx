import { fireEvent, render, screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import { store } from '../../../src/config/redux/store';
import { Counter } from '../../../src/features/counter/Counter';

describe('tests in <Counter />', () => {
  const initialValue = 100;

  test('should match with the snapshot', () => {
    const { container } = render(
      <Provider store={store}>
        <Counter value={initialValue} />
      </Provider>
    );
    expect(container).toMatchSnapshot();
  });

  test('should show the initial value 100', () => {
    render(
      <Provider store={store}>
        <Counter value={initialValue} />
      </Provider>
    );
    expect(screen.getByText(initialValue)).toBeTruthy();
  });

  test('should increment in 1', () => {
    render(
      <Provider store={store}>
        <Counter value={initialValue} />
      </Provider>
    );
    const btnIncrement = screen.getByRole('button', { name: 'btn-increment' });
    fireEvent.click(btnIncrement);
    expect(screen.getByText(initialValue + 1)).toBeTruthy();
  });

  test('should decrement in 1', () => {
    render(
      <Provider store={store}>
        <Counter value={initialValue} />
      </Provider>
    );
    const btnDecrement = screen.getByRole('button', { name: 'btn-decrement' });
    fireEvent.click(btnDecrement);
    expect(screen.getByText(initialValue - 1)).toBeTruthy();
  });

  test('should reset to prop value', () => {
    render(
      <Provider store={store}>
        <Counter value={initialValue} />
      </Provider>
    );
    const btnIncrement = screen.getByRole('button', { name: 'btn-increment' });
    const btnReset = screen.getByRole('button', { name: 'btn-reset' });
    fireEvent.click(btnIncrement);
    fireEvent.click(btnIncrement);
    fireEvent.click(btnIncrement);
    fireEvent.click(btnReset);
    expect(screen.getByText(initialValue)).toBeTruthy();
  });
});
