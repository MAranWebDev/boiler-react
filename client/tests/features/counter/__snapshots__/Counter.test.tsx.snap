// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests in <Counter /> should match with the snapshot 1`] = `
<div>
  <div>
    <section>
      <h2>
        100
      </h2>
    </section>
    <section>
      <button
        aria-label="btn-increment"
        type="button"
      >
        +1
      </button>
      <button
        aria-label="btn-decrement"
        type="button"
      >
        -1
      </button>
      <button
        aria-label="btn-reset"
        type="button"
      >
        Reset
      </button>
    </section>
  </div>
</div>
`;
